modules/jfx.incubator.richtext/src/main/java/jfx/incubator/scene/control/richtext/RichTextArea.java

```diff
@@ -0,0 +1,2396 @@
+     */
+    public void applyStyle(TextPos start, TextPos end, StyleAttributeMap attrs) {
+        StyledTextModel m = getModel();
+        m.applyStyle(start, end, attrs, true);
+    }
+
+    /**
+     * When selection exists, deletes selected text.  Otherwise, deletes the character preceding the caret,
+     * possibly breaking up the grapheme clusters.
+     * This method does nothing if either control or the model is not editable, or the caret position is {@code null}.
+     * <p>
+     * This action can be changed by remapping the default behavior via {@link InputMap}.
+     *
+     * @see RichTextArea.Tags#BACKSPACE
+     */
+    public void backspace() {
+        execute(Tags.BACKSPACE);
+    }
+
+    /**
+     * Clears the document, creating an undo entry.
+     *
+     * @throws NullPointerException if the model is {@code null}
+     * @throws UnsupportedOperationException if the model is not {@link StyledTextModel#isWritable() writable}
+     */
+    public final void clear() {
+        TextPos end = getDocumentEnd();
+        replaceText(TextPos.ZERO, end, StyledInput.EMPTY, true);
+    }
+
+    /**
+     * Clears existing selection, if any.  This method is an alias for {@code getSelectionModel().clear()}.
+     */
+    public final void clearSelection() {
+        selectionModel.clear();
+    }
+
+    /**
+     * Clears the undo-redo stack of the underlying model.
+     * This method does nothing if the model is null.
+     */
+    public final void clearUndoRedo() {
+        StyledTextModel m = getModel();
+        if (m != null) {
+            m.clearUndoRedo();
+        }
+    }
+
+    /**
+     * When selection exists, copies the selected rich text to the clipboard in all the formats supported
+     * by the model.
+     * <p>
+     * This method does nothing when the caret position is {@code null}.
+     * <p>
+     * This action can be changed by remapping the default behavior via {@link InputMap}.
+     * @see RichTextArea.Tags#COPY
+     */
+    public void copy() {
+        execute(Tags.COPY);
+    }
+
+    /**
+     * Copies the selected text in the specified format to the clipboard.
+     * This method does nothing if no selection exists or when the data format is not supported by the model.
+     *
+     * @param format the data format to use
+     */
+    public final void copy(DataFormat format) {
+        RichTextAreaSkin skin = richTextAreaSkin();
+        if (skin != null) {
+            skin.copyText(format);
+        }
+    }
+
+    /**
+     * Transfers the currently selected text to the clipboard,
+     * removing the current selection.
+     * <p>
+     * This method does nothing if the model or the caret position is {@code null}.
+     * The default implementation may throw an {@code UnsupportedOperationException}
+     * if the control is not editable.  When this happens, the copying to clipboard will succeed.
```

Comment:
the javadoc was incorrect, thanks for catching.
rephrased to say
"Any failures, such as the control being not editable, or out of memory condition, will be silently ignored. "

Currently, there is no way to provide an error feedback (audio or visual), such as `beep()`.  Perhaps we need to add Tags.ERROR so that application can implement its own.

Another issue is logging - it would be nice to provide a mechanism to plug in into application logging subsystem, whatever that is.  create a protected method?

Reply:
I don't think we want to get into the logging business. We currently use the System logger for all of our logging. Unless there is some compelling reason we need to do more, we should stick with that.

Reply:
Right.

I think providing error feedback via new `Tags.ERROR_FEEDBACK` should be sufficient.  The default is to do nothing (javafx does not have a `beep()`) but the applications are free to supply their own implementation.